File: go/algorithms/cycle.go
package algorithms

// CycleSort minimizes writes by rotating cycles; sorts in-place.
func CycleSort(arr []int) {
    n := len(arr)
    for cycleStart := 0; cycleStart+1 < n; cycleStart++ {
        item := arr[cycleStart]
        pos := cycleStart
        for i := cycleStart + 1; i < n; i++ {
            if arr[i] < item {
                pos++
            }
        }
        if pos == cycleStart {
            continue
        }
        for pos < n && item == arr[pos] {
            pos++
        }
        if pos < n {
            arr[pos], item = item, arr[pos]
        }
        for pos != cycleStart {
            pos = cycleStart
            for i := cycleStart + 1; i < n; i++ {
                if arr[i] < item {
                    pos++
                }
            }
            for pos < n && item == arr[pos] {
                pos++
            }
            if pos < n {
                arr[pos], item = item, arr[pos]
            }
        }
    }
}

File: rust/src/cycle.rs
pub fn cycle_sort(arr: &mut [i32]) {
    let n = arr.len();
    for cycle_start in 0..n.saturating_sub(1) {
        let mut item = arr[cycle_start];
        let mut pos = cycle_start;
        for i in cycle_start + 1..n {
            if arr[i] < item {
                pos += 1;
            }
        }
        if pos == cycle_start {
            continue;
        }
        while pos < n && item == arr[pos] {
            pos += 1;
        }
        if pos < n {
            std::mem::swap(&mut arr[pos], &mut item);
        }
        while pos != cycle_start {
            pos = cycle_start;
            for i in cycle_start + 1..n {
                if arr[i] < item {
                    pos += 1;
                }
            }
            while pos < n && item == arr[pos] {
                pos += 1;
            }
            if pos < n {
                std::mem::swap(&mut arr[pos], &mut item);
            }
        }
    }
}
